### PR TITLE
Bug 709348 - rework storing of records, track stored records to prevent reupload

### PR DESCRIPTION
**Pull request by rnewman**
This hasn't been tested on a real device, but it does pass automated tests (as well as a new one), and broke tests in the expected places :)

Input requested. Note the long sequence of commits that includes some evolutionary steps for clarity.

Nick, you'll notice a bunch of commits here for moving WBORepository. I hope that doesn't spike you too much!

### CODE REVIEW

**Comment by ncalexan** [comment]
Two general comments that I also brought up on IRC.

I find the term 'tracking' to be really ambiguous.  You could track records for a million reasons, only one of which is not reuploading.  The crucial method is 'isTrackedForExclusion' suggests trackedForExclusion throughout.

I also have an ideological issue with testing things in the callback.  If your callbacks never get called, nothing will happen, but your tests won't fail.  I prefer having the callbacks collect things which are then verified after the action is over.  Then you really know that things happened.  And you can also check that bad things _didn't_ happen.  It's also a lot easier to deduce what you wanted to happen -- rather than reverse engineering when the callbacks are called, you're reading someone assert something about the state of a callback object.

**Comment by rnewman** [comment]
Re testing in callbacks: if the expected callback method isn't called, then it also won't notify the monitor, and the test won't complete. (And the other arms of the delegate all fail noisily by default.)

There's no other part of the code that calls performNotify with the appropriate test helper, so if the test completes it's because it's passed through our delegate methods.

If our tests are passing because we're not waiting for the callback to be called, then we have bigger problems than testing prior to notifying.

I understand your point, and I agree to an extent, but it's often more convenient to do checks where the objects are available for inspection.

File: src/main/java/org/mozilla/gecko/sync/repositories/HashSetStoreTracker.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  // Used to store GUIDs that were not locally modified but
+  // have been modified by a call to `store`, and thus
+  // should not be returned by a subsequent fetch.
+  private HashSet<String> guids;
+
+  public HashSetStoreTracker() {
+    guids = new HashSet<String>();
+  }
+
+  public String toString() {
```
Comment (liuche) [inline src/main/java/org/mozilla/gecko/sync/repositories/HashSetStoreTracker.java:24]: @Override annotation.

File: src/main/java/org/mozilla/gecko/sync/repositories/RepositorySession.java [hunk 2]
```diff
@@ -251,4 +256,91 @@ public void abort() {
...
+  protected Record reconcileRecords(final Record remoteRecord,
+                                    final Record localRecord,
+                                    final long lastRemoteRetrieval,
+                                    final long lastLocalRetrieval) {
+    Log.d(LOG_TAG, "Reconciling remote " + remoteRecord.guid + " against local " + localRecord.guid);
+
+    if (localRecord.equalPayloads(remoteRecord)) {
+      if (remoteRecord.lastModified > localRecord.lastModified) {
+        Log.d(LOG_TAG, "Records are equal. No record application needed.");
+        return null;
```
Comment (liuche) [inline src/main/java/org/mozilla/gecko/sync/repositories/RepositorySession.java:304]: Nit: Both nested if checks return null, can remove this return and just fall through, unless this is for comment clarity.

Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/repositories/RepositorySession.java:304]: Yeah, it's for clarity. I want all this logic to be as obvious as possible.

(I already had one bug around fallthrough in WBORepository that my tests caught.)

File: test/src/org/mozilla/android/sync/test/StoreTrackingTest.java [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+      // Should not happen.
+    }
+  }
+
+  private void doTestNewSessionRetrieveByTime(final WBORepository repository,
+                                              final String expectedGUID) {
+    final SuccessCreationDelegate createDelegate = new SuccessCreationDelegate() {
+      @Override
+      public void onSessionCreated(final RepositorySession session) {
+        System.out.println("Session created.");
```
Comment (liuche) [inline test/src/org/mozilla/android/sync/test/StoreTrackingTest.java]: Use Log instead? Not sure about when we want to use System.out.println.

Comment (rnewman) [inline test/src/org/mozilla/android/sync/test/StoreTrackingTest.java]: Yeah, this is test code, so it doesn't really matter, but I should be consistent.

**Comment by ncalexan** [comment]
Looks good to me.

**Pull request closed without merging**